File: Graph/Leetcode/Medium/GraphValidTree/GraphValidTree.cpp
// https://neetcode.io/problems/valid-tree
// https://leetcode.com/problems/graph-valid-tree/description/

#include<bits/stdc++.h>
using namespace std;
bool dfs(vector<vector<int>> &adj, vector<bool> &visited, int curr, int prev)
{
    if (visited[curr])
    {
        return true;
    }

    visited[curr] = true;
    for (int &neighbor : adj[curr])
    {
        if (neighbor != prev && dfs(adj, visited, neighbor, curr))
        {
            return true;
        }
    }
    return false;
}
bool validTree(int n, vector<vector<int>> &edges)
{
    vector<bool> visited(n, false);
    vector<vector<int>> adj(n);

    for (auto &edge : edges)
    {
        adj[edge[0]].emplace_back(edge[1]);
        adj[edge[1]].emplace_back(edge[0]);
    }

    if (dfs(adj, visited, 0, -1))
        return false;

    for (int i = 1; i < n; i++)
    {
        if (!visited[i])
            return false;
    }

    return true;
}

File: Graph/Leetcode/Medium/GraphValidTree/GraphValidTree.java
// https://neetcode.io/problems/valid-tree
// https://leetcode.com/problems/graph-valid-tree/description/

import java.util.ArrayList;
import java.util.List;

public class GraphValidTree {
    boolean dfs(List<List<Integer>> adj, boolean[] visited, int curr, int prev) {
        if(visited[curr]) {
            return true;
        }

        visited[curr] = true;

        for(int neighbor: adj.get(curr)) {
            if(neighbor != prev && dfs(adj, visited, neighbor, curr))
                return true;
        }

        return false;
    }
    public boolean validTree(int n, int[][] edges) {
        List<List<Integer>> adj = new ArrayList<>();
        for(int i=0; i<n; i++) {
            adj.add(new ArrayList<>());
        }

        for(int[] edge: edges) {
            adj.get(edge[0]).add(edge[1]);
            adj.get(edge[1]).add(edge[0]);
        }

        boolean[] visited = new boolean[n];

        if(dfs(adj, visited, 0, -1)) return false;

        for(int i=1; i<n; i++) {
            if(!visited[i]) return false;
        }

        return true;
    }
}
